csharp/src/Ice/Communicator.cs

```diff
@@ -222,15 +242,19 @@ internal CancellationToken CancellationToken
             = new ThreadLocal<SortedDictionary<string, string>>();
         private volatile IReadOnlyDictionary<string, string> _defaultContext =
             ImmutableSortedDictionary<string, string>.Empty;
+        private volatile ImmutableList<InvocationInterceptor> _defaultInvocationInterceptors =
+            ImmutableList<InvocationInterceptor>.Empty;
         private volatile ILocatorPrx? _defaultLocator;
         private volatile IRouterPrx? _defaultRouter;
-        private readonly List<DispatchInterceptor> _dispatchInterceptors = new List<DispatchInterceptor>();
+        private volatile ImmutableList<DispatchInterceptor> _defaultDispatchInterceptors =
+            ImmutableList<DispatchInterceptor>.Empty;
         private Task? _disposeTask;
-        private readonly List<InvocationInterceptor> _invocationInterceptors = new List<InvocationInterceptor>();
         private static readonly Dictionary<string, Assembly> _loadedAssemblies = new Dictionary<string, Assembly>();
         private readonly ConcurrentDictionary<ILocatorPrx, LocatorInfo> _locatorInfoMap =
             new ConcurrentDictionary<ILocatorPrx, LocatorInfo>();
 
+        private volatile ILogger _logger;
```

Comment:
The communicator's logger is accessed quite often by the code. Doesn't the volatile here has a cost and do we really need it?

Reply:
I had the same concern, but I think removing volatile would be a premature optimization at the cost of correctness.
Also, logger should not be use that often: for tracing, we always check the trace category before using the logger.